Configurable variants saving to cache OLD

bricht opened the pull request:
-Module and changes to existing infrastructure to support product variants
-Added configurable product variants
-Added tests for product variants

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

### Related Pull Requests
<!-- related pull request placeholder -->
<!--- ^ DO NOT REMOVE placeholder above ^ -->



### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/catalog-storefront#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Code Review Checklist (*)


- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
- [ ] no backward incompatible changes

Commit:
Product variants
-Module and changes to existing infrastructure to support product variants
-Added configurable product variants
-Added tests for product variants

Commit:
Product variants
Minor syntax fixes

Commit:
Product variants
-Changed entity type to string

Commit:
Product variants
-Test fixes

Commit:
Product variants
-Minor fixes

Commit:
Product variants
-Moved test from variants to configurable variants and other minor adjustments

Commit:
Product variants

Commit:
Product variants

bricht commented:
@magento run all tests
without extensions magento/saas-export, magento/data-solutions-services-id, magento/magento-services-connector

RuslanKostiv1 commented on app/code/Magento/CatalogExportApi/Api/ProductVariantRepositoryInterface.php:
```diff
@@ -0,0 +1,30 @@
+     * @param string[] $ids
+     * @return \Magento\CatalogExportApi\Api\Data\ProductVariant[]
+     */
+    public function get(array $ids): array;
```
where this API is used?

bricht replied:
\Magento\CatalogMessageBroker\Model\FetchProductVariants::execute

The getByProductIds is currently not really used, however I think we will likely need it in future so I added it.

RuslanKostiv1 commented on app/code/Magento/ConfigurableProductDataExporter/Model/Provider/Product/ProductVariants/ConfigurableId.php:
```diff
@@ -0,0 +1,29 @@
+     * @param string[] $params
+     * @return string
+     */
+    public function resolve(string ...$params): string
+    {
+        array_unshift($params, ConfigurableOptionValueUid::OPTION_TYPE);
+        return implode('/', $params);
```
propose to change a little bit this approach to make is more safe, because in this case we do not know nothing about order of incoming parameters

```
// \Magento\ConfigurableProductDataExporter\Model\Provider\Product\ProductVariants::getVariants
$id = $idResolver->resolve(\...(['parentId' => $row['parentId'], 'childId' => $row['childId']));
....
```
in this case we can control data:
```
$params = [
   'parentId' => $row['parentId'],
   'childId' => $row['childId']
];
```

bricht replied:
Done

RuslanKostiv1 commented on app/code/Magento/ConfigurableProductDataExporter/Model/Provider/Product/Variants.php:
```diff
@@ -7,13 +7,14 @@
 
 /**
  * Configurable product variant data provider
+ * TODO: Deprecated. Remove this class and its query class.
```
please add id of the issue here

bricht replied:
Done

Commit:
Product variants
-Comment fix
-Made product variant id provider params more defined

Commit:
Product variants
-Removed indexation reliance on product index
-Changed mview to product_relations

Commit:
Product variants
-Fixed indexing issues

Commit:
Product variants

Commit:
Merge branch 'main' into 27_product-variants
